### PR TITLE
8282590: C2: assert(addp->is_AddP() && addp->outcnt() > 0) failed: Don't process dead nodes

### PR DESCRIPTION
**Pull request by eme64**
Problem:
Sometimes nodes are generated, and then not properly added to any other node, nor added to the worklist, so that it could be deleted. In some cases, a node can thus survive after IGVN and enter into `ConnectionGraph::compute_escape`, where no dead nodes are expected (except constants).

In `ArrayCopyNode::prepare_array_copy`:
In one, I now add the new ConvI2LNode to the worklist if we are aborting.
In the other case, I could safely reorder the code, such new ConvI2LNode would only be generated once we know we are not going to abort.

In `SubTypeCheckNode::Ideal`:
Replaced `set_req` with `set_req_X`, such that the input-node that is replaced (which is now potentially a dead node) is put on the worklist.

In `ArrayCopyNode::Ideal`:
If we decide to `return NULL` we also need to add the potentially dead forward / backward memory nodes to the worklist, they may turn out to be dead.

Added a regression test: the test + flag combination that first reproduced this issue.

All Tests, including this new regression test are passing. Tests rerun after small refactorings.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282590](https://bugs.openjdk.java.net/browse/JDK-8282590): C2: assert(addp->is_AddP() && addp->outcnt() > 0) failed: Don't process dead nodes

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7728/head:pull/7728` \
`$ git checkout pull/7728`

Update a local copy of the PR: \
`$ git checkout pull/7728` \
`$ git pull https://git.openjdk.java.net/jdk pull/7728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7728`

View PR using the GUI difftool: \
`$ git pr show -t 7728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7728.diff">https://git.openjdk.java.net/jdk/pull/7728.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/arraycopynode.cpp [hunk 1]
```diff
@@ -629,6 +641,15 @@ Node *ArrayCopyNode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
   if (!finish_transform(phase, can_reshape, ctl, mem)) {
+    if (can_reshape) {
+      // make sure the just created clones are taken care of if they now are dead
+      if (!forward_mem->is_top() && forward_mem->outcnt() == 0) {
+        phase->is_IterGVN()->_worklist.push(forward_mem);
+      }
+      if (!backward_mem->is_top() && backward_mem->outcnt() == 0) {
+        phase->is_IterGVN()->_worklist.push(backward_mem);
+      }
+    }
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/arraycopynode.cpp:652]: I don't think you need to check if the mem is really dead. Adding it to the worklist does not hurt. Would this work as well?

```
if (can_reshape) {
  phase->is_IterGVN()->_worklist.push(mem):
}
```

File: src/hotspot/share/opto/arraycopynode.cpp [hunk 2]
```diff
@@ -294,9 +294,20 @@ bool ArrayCopyNode::prepare_array_copy(PhaseGVN *phase, bool can_reshape,
...
+      // Offset is out of bounds (the ArrayCopyNode will be removed)
+      return false;
+    }
     dest_offset = Compile::conv_I2X_index(phase, dest_offset, ary_dest->size());
-    if (src_offset->is_top() || dest_offset->is_top()) {
+    if (dest_offset->is_top()) {
       // Offset is out of bounds (the ArrayCopyNode will be removed)
+      PhaseIterGVN* igvn = phase->is_IterGVN();
+      if (can_reshape) {
+        if (!src_offset->is_top()) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/arraycopynode.cpp:306]: The ifs can be merged.

Comment (TobiHartmann) [reply src/hotspot/share/opto/arraycopynode.cpp:306]: And looking at this again, `src_offset` can't be top, right? We just checked it above.

**Review by TobiHartmann — APPROVED** [review]
Looks good, thanks for making these changes.

**Comment by eme64** [comment]
Thanks @TobiHartmann and @chhagedorn for the help, suggestions for improvement and reviews!
/integrate

**Pull request closed without merging**